### PR TITLE
Stop retry for unauthorized errors

### PR DESCRIPTION
**Pull request by duanxianbo**

### CODE REVIEW

File: src/libs/retry/infinite.test.ts [hunk 1]
```diff
@@ -1,14 +1,14 @@
...
-import { Errors, InternalError } from '@lib/error/errors';
 import { TestRandomNumberGenerator } from '@lib/math/randomNumberGenerator';
 import { TestRuntime } from '@lib/runtime/runtime';
 import { Semaphore } from '@lib/sync/semaphore';
 
 import { ExponentialBackoff } from './backoff/exponential';
 import { InfiniteRetry } from './infinite';
-import { Executable } from './retry.test';
+import { Executable } from './retry';
+
```
Comment (naturetime10) [inline src/libs/retry/infinite.test.ts:12]:
```suggestion
```

File: src/libs/retry/infinite.ts [hunk 2]
```diff
@@ -19,10 +24,18 @@ export class InfiniteRetry implements Retry {
                     resolve([retries, undefined]);
                     return;
                 }
+                retries++;
```
Comment (naturetime10) [inline src/libs/retry/infinite.ts:27]:
```suggestion

                retries++;
```

File: src/libs/retry/infinite.ts [hunk 3]
```diff
@@ -19,10 +24,18 @@ export class InfiniteRetry implements Retry {
...
+                const errorCategory = ERROR_CATEGORIES[err.errorCode];
+                if (errorCategory === 'ClientInteraction') {
+                    await this.beforeThreadSleep?.call(null)
```
Comment (naturetime10) [inline src/libs/retry/infinite.ts]: Can we add `beforeSkipRetry`  instead of using `beforeThreadSleep`?

File: src/libs/retry/infinite.ts [hunk 4]
```diff
@@ -19,10 +24,18 @@ export class InfiniteRetry implements Retry {
...
+                    resolve([retries, err]);
+                    return;
+                }
 
                 this.backoff.onFailure();
-                await this.runtime.threadSleep(this.backoff.delay());
-                retries++;
+                const duration = this.backoff.delay()
+                await this.beforeThreadSleep?.call(null, duration)
+                await this.runtime.threadSleep(duration);
```
Comment (naturetime10) [inline src/libs/retry/infinite.ts]: Undo this change.

File: src/libs/retry/infinite.ts [hunk 5]
```diff
@@ -1,11 +1,16 @@
...
 
 import { BackOff } from './backoff/backoff';
 import { Retry } from './retry';
 
 export class InfiniteRetry implements Retry {
-    constructor(private runtime: Runtime, private backoff: BackOff) {}
+    constructor(
+        private runtime: Runtime,
+        private backoff: BackOff,
+        private beforeThreadSleep?: (duration?: Duration) => Promise<void>,
```
Comment (naturetime10) [inline src/libs/retry/infinite.ts]:
```suggestion
        private beforeSkipRetry?: () => Promise<void>,
```

**Review by naturetime10 — CHANGES_REQUESTED** [review]

File: src/libs/retry/infinite.ts (hunk 3 above)
Comment (duanxianbo) [reply src/libs/retry/infinite.ts]: yes, so you are saying pass another optional param beforeSkipRetry

File: src/libs/retry/infinite.test.ts [hunk 6]
```diff
@@ -26,12 +26,16 @@ describe('InfiniteRetry', () => {
...
+                    return {
+                        errorCode: 'RequestTimeOut',
+                        message: 'first error'
+                    }
                 }
             },
         };
         beforeSleepSemaphore = new Semaphore('BeforeSleep');
         exitSemaphore = new Semaphore('Exit');
+        beforeSkipRetry = async () => {beforeSleepSemaphore.release()}
```
Comment (naturetime10) [inline src/libs/retry/infinite.test.ts]:
```suggestion
        beforeSkipRetry = async () => {
            beforeSleepSemaphore.release();
        }
```

File: src/libs/retry/infinite.test.ts [hunk 7]
```diff
@@ -17,6 +16,7 @@ describe('InfiniteRetry', () => {
     let beforeSleepSemaphore: Semaphore;
```
Comment (naturetime10) [inline src/libs/retry/infinite.test.ts:19]:
```suggestion
    let beforeRetrySemaphore: Semaphore;
```

File: src/libs/retry/maxCount.test.ts [hunk 8]
```diff
@@ -17,82 +17,80 @@ describe('MaxCount', () => {
...
-            const [retries, error] = await maxCount.withRetry(
+            const [retriesRes, errorRes] = await maxCount.withRetry(
                 executable.execute.bind(executable),
             );
 
-            strictEqual(retries, 2);
+            strictEqual(retriesRes, retries);
             strictEqual(duration.totalMilliSeconds, 8);
-            strictEqual(error?.message, 'first error');
+            deepStrictEqual(errorRes, error);
```
Comment (naturetime10) [inline src/libs/retry/maxCount.test.ts]: How about only compare `errorCode`. Error message is only for human to read and is subject to change in the future?

File: src/libs/retry/maxCount.ts [hunk 9]
```diff
@@ -17,17 +18,24 @@ export class MaxCountRetry implements Retry {
...
             let err: InternalError | undefined;
+
             while (retries < this.maxCount) {
                 err = await execute();
                 if (err === undefined) {
                     this.backoff.onSuccess();
                     resolve([retries, undefined]);
                     return;
                 }
+                retries++
```
Comment (naturetime10) [inline src/libs/retry/maxCount.ts]:
```suggestion

                retries++
```

File: src/libs/retry/retry.ts [hunk 10]
```diff
@@ -5,3 +5,8 @@ export interface Retry {
         execute: () => Promise<InternalError | undefined>,
     ) => Promise<[number, InternalError | undefined]>;
 }
+
+export interface Executable {
+    count: number,
+    execute: () => Promise<InternalError | undefined>,
+}
```
Comment (naturetime10) [inline src/libs/retry/retry.ts]: Are we using it in production or testing code only? If only for testing code, can we keep it in test modules?

File: src/libs/retry/maxCount.ts (hunk 9 above)
Comment (naturetime10) [reply src/libs/retry/maxCount.ts]: yarn format

**Review by naturetime10 — CHANGES_REQUESTED** [review]

File: src/libs/retry/maxCount.test.ts (hunk 8 above)
Comment (duanxianbo) [reply src/libs/retry/maxCount.test.ts]: yes

File: src/libs/retry/infinite.test.ts [hunk 11]
```diff
@@ -1,67 +1,117 @@
...
 import { Semaphore } from '@lib/sync/semaphore';
 
 import { ExponentialBackoff } from './backoff/exponential';
 import { InfiniteRetry } from './infinite';
-import { Executable } from './retry.test';
+
+export interface Executable {
+    count: number;
+    execute: () => Promise<InternalError | undefined>;
+}
```
Comment (naturetime10) [inline src/libs/retry/infinite.test.ts]: test-utils.ts

**Review by naturetime10 — APPROVED** [review]

File: src/libs/retry/infinite.test.ts (hunk 11 above)
Comment (naturetime10) [reply src/libs/retry/infinite.test.ts]: Please make this change and merge afterwards.

**Pull request merged**